File: packages/java-pom-bdd/src/test/java/com/example/pages/CatalogPage.java
package com.example.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class CatalogPage extends BasePage {
    private final Locator searchInput;
    private final Locator sortSelect;
    private final Locator categoryTabs;
    private final Locator productCards;

    public CatalogPage(Page page) {
        super(page);
        this.searchInput = page.getByPlaceholder("商品を検索...");
        this.sortSelect = page.getByRole(com.microsoft.playwright.options.AriaRole.COMBOBOX,
                new Page.GetByRoleOptions().setName("並び替え"));
        this.categoryTabs = page.getByRole(com.microsoft.playwright.options.AriaRole.TABLIST,
                new Page.GetByRoleOptions().setName("商品カテゴリ"));
        this.productCards = page.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
                new Page.GetByRoleOptions().setName(Pattern.compile("カートに追加")));
    }

    public void searchProduct(String keyword) {
        searchInput.fill(keyword);
        page.waitForTimeout(500);
    }

    public void clearSearch() {
        searchInput.clear();
    }

    public void sortBy(String option) {
        sortSelect.selectOption(option);
    }

    public void filterByCategory(String category) {
        categoryTabs.getByRole(com.microsoft.playwright.options.AriaRole.TAB,
                new Locator.GetByRoleOptions().setName(category)).click();
    }

    public int getProductCount() {
        return productCards.count();
    }

    public void addToCart(String productName) {
        Locator productImg = page.getByRole(com.microsoft.playwright.options.AriaRole.IMG,
                new Page.GetByRoleOptions().setName(productName));
        Locator card = productImg.locator("..");
        card.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
                new Locator.GetByRoleOptions().setName(Pattern.compile("カートに追加"))).click();
    }

    public boolean isAddToCartEnabled(String productName) {
        Locator productImg = page.getByRole(com.microsoft.playwright.options.AriaRole.IMG,
                new Page.GetByRoleOptions().setName(productName));
        Locator card = productImg.locator("..");
        Locator button = card.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
                new Locator.GetByRoleOptions().setName(Pattern.compile("カートに追加")));
        return button.isEnabled();
    }

    public boolean isProductVisible(String productName) {
        return page.locator("text=" + productName).isVisible();
    }

    public void clickCategoryTab(String category) {
        categoryTabs.getByRole(com.microsoft.playwright.options.AriaRole.TAB,
                new Locator.GetByRoleOptions().setName(category)).click();
    }

    public boolean areCategoryTabsVisible() {
        return categoryTabs.isVisible();
    }

    public boolean isSortedByPriceAsc() {
        List<Locator> cards = page.locator("[class*='MuiCard-root']").all();
        if (cards.size() < 2) return true;

        List<Integer> prices = new ArrayList<>();
        for (Locator card : cards) {
            String priceText = card.locator("text=/¥[0-9,]+/").first().textContent();
            if (priceText != null) {
                String numStr = priceText.replace("¥", "").replace(",", "");
                prices.add(Integer.parseInt(numStr));
            }
        }

        for (int i = 1; i < prices.size(); i++) {
            if (prices.get(i) < prices.get(i - 1)) {
                return false;
            }
        }
        return true;
    }
}

File: packages/ts-bdd-soft-assertion/pages/CatalogPage.ts
import { Page, Locator } from '@playwright/test';

export class CatalogPage {
  private readonly searchInput: Locator;
  private readonly sortSelect: Locator;
  private readonly categoryTabs: Locator;
  private readonly productCards: Locator;

  constructor(private page: Page) {
    this.searchInput = page.getByPlaceholder('商品を検索...');
    this.sortSelect = page.getByRole('combobox', { name: '並び替え' });
    this.categoryTabs = page.getByRole('tablist', { name: '商品カテゴリ' });
    // Product cards are containers that have "カートに追加" button
    this.productCards = page.getByRole('button', { name: /カートに追加/ });
  }

  async searchProduct(keyword: string) {
    await this.searchInput.fill(keyword);
    await this.page.waitForTimeout(500); // Wait for filter to apply
  }

  async clearSearch() {
    await this.searchInput.clear();
  }

  async sortBy(option: '名前順' | '価格（安い順）' | '価格（高い順）') {
    await this.sortSelect.selectOption(option);
  }

  async filterByCategory(category: 'すべての商品' | '電子機器' | '衣類' | '書籍' | 'ホーム') {
    await this.categoryTabs.getByRole('tab', { name: category }).click();
  }

  async getProductCount() {
    return await this.productCards.count();
  }

  async getProductNames() {
    const names: string[] = [];
    const count = await this.productCards.count();
    for (let i = 0; i < count; i++) {
      const card = this.productCards.nth(i);
      const nameElement = card.locator('text=/^[^¥]+$/').first();
      const name = await nameElement.textContent();
      if (name) names.push(name.trim());
    }
    return names;
  }

  async addToCart(productName: string) {
    // Use img alt attribute to locate the product card
    const productImg = this.page.getByRole('img', { name: productName });
    const card = productImg.locator('..');
    await card.getByRole('button', { name: /カートに追加/i }).click();
  }

  async isAddToCartEnabled(productName: string) {
    const productImg = this.page.getByRole('img', { name: productName });
    const card = productImg.locator('..');
    const button = card.getByRole('button', { name: /カートに追加/i });
    return await button.isEnabled();
  }

  async getProductPrice(productName: string) {
    const productImg = this.page.getByRole('img', { name: productName });
    const card = productImg.locator('..');
    const priceText = await card.locator('text=/¥[\\d,]+/').textContent();
    if (priceText) {
      const match = priceText.match(/¥([\d,]+)/);
      if (match) {
        return parseInt(match[1].replace(',', ''), 10);
      }
    }
    return 0;
  }

  async isProductVisible(productName: string) {
    return await this.page.locator(`text=${productName}`).isVisible();
  }
}
